src/AsyncTestCase.php

```diff
@@ -0,0 +1,34 @@
+<?php declare(strict_types=1);
+
+/**
+ * A PHPUnit TestCase intended to help facilitate writing async tests by running each test on the amphp Loop and
+ * ensuring that the test runs until completion based on your test returning either a Promise or a Generator.
+ */
+
+namespace Amp\PHPUnit;
+
+use Amp\Loop;
+use PHPUnit\Framework\TestCase as PHPUnitTestCase;
+use function Amp\call;
+
+abstract class AsyncTestCase extends PHPUnitTestCase {
+
+    public function runTest() {
+        Loop::run(function() {
+            $testTimeout = $this->getTestTimeout();
+            $watcherId = Loop::delay($testTimeout, function() use($testTimeout) {
+                Loop::stop();
+                $this->fail('Expected test to complete before ' . $testTimeout . 'ms time limit');
+            });
+            yield call(function() {
+                return parent::runTest();
+            });
+            Loop::cancel($watcherId);
+        });
+    }
+
+    protected function getTestTimeout() : int {
+        return 1500;
+    }
```

Comment:
There's no need to couple the timeout to the test run IMO. If you need a timeout, that should be explicit, e.g. `$this->timeout(1500);` within a test case that fails the test after 1500ms. What do you think about that?

Reply:
I guess I consider a test that runs for too long a bug or a code smell. And I consider an individual test running for more than 1.5 seconds is too long. I think the default case should be to expect your test to complete within that time and being explicit only if your test needs to run longer.

However, I'm not necessarily sold on this opinion and could accomplish my use case by setting the timeout in a `setUp` function.